### PR TITLE
Miniature Golf

### PR DESCRIPTION
**Pull request by Schmitz-David**
Better late then never...

### CODE REVIEW

**Comment by Shoukox** [comment]
# Review of First Submission (Miniature Golf)
## 1. Similarity to ALE: 3.75/4
The implementation appears to be almost identical to the ALE version.
However, there is only one noticable difference: the background colors are static. In several Youtube videos about this game, different background colors are used for each level.
## 2. JAX-ness and quality of the code: 4/4
The naming is clear and easy to understand. The code is readable and well structured. In addition, many helper functions are implemented, which further improves the overall readability and maintainability of the code. The implementation is also JAX-compatible.

**Comment by Schmitz-David** [comment]
@Shoukox Ah yes, at the start and end of the game, the screen is darker. The game is in "select mode", where for the Atari 2600 it was possible to use the "Game Select" switch to select a difficulty (if I recall correctly, "Pro" has a faster moving ball than "Amateur"). Thus, it's not really part of the gameplay and since we should imitate the ale_py versions (which does not even have the "Game Select" action), I did not implement this.

**Comment by Schmitz-David** [comment]
@PaulSeitz A tangentially related question: can we somehow easily profile the performance of our code? I have tried training a PPN agent (without any specialized GPU) and training for 1e4 steps takes 6 minutes. At this rate, I would need to wait for days for the training to finish with the standard parameters (5e7 steps). Of course, using a good GPU would increase the available parallelism, but I still suspect that something is off. I have tested jax.profile.trace before (with the Perfetto UI), but it doesn't seem suitable for the jitted training process.

**Comment by Schmitz-David** [comment]
@PaulSeitz Despite the non-ideal performance, I would still leave the reward function as is and evaluate the different agents on each mod using the current setup. Is this fine? This should be sufficiently interesting given that the two return curves above differ quite substantially and it's also nice to have games where the current agents do not train particularly well, as these can serve as a benchmark to test future improvements.

**Comment by Schmitz-David** [comment]
Alright, so I had to revert many files back to an ancient state for the tests to pass, but the PR should now be ready for submission. Regarding the change to flax datastructures, the ["updated" branch](https://github.com/Schmitz-David/JAXAtari/tree/updated) contains these changes. However, since there are still some failing tests, I did not want to include this for the final submission.

**Comment by Schmitz-David** [comment]
Okay, it seems to me that the previous problems with migration to flax.PyTreeNode are overcome, so I added the updates to the master branch. In case something is not correct with these newer changes, rolling back this last commit should suffice. In any case, this is now the final submission.
